ets-demo/etsdemo/loader.py

```diff
@@ -0,0 +1,121 @@
+# (C) Copyright 2020 Enthought, Inc., Austin, TX
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+
+""" Module for loading example data files contributed by other distributions
+via entry points.
+"""
+
+import logging
+import os
+
+import pkg_resources
+
+from etsdemo.app import DemoPath, DemoVirtualDirectory
+
+
+logger = logging.getLogger(__name__)
+
+#: The entry point group name for other packages to contribute
+#: data files.
+ENTRY_POINT_GROUP = "etsdemo_data"
+
+
+def get_responses():
+    """ Load entry points and return the responses.
+
+    Version 1 response should have the following format:
+    {
+        "version": 1,
+        "name": <str>,
+        "root": <str>,
+    }
+
+    Required keys:
+    ``version``: int
+        The version for the format of this response
+    ``name``: str
+        The main display name for the group of resources.
+    ``root``: str
+        The root directory where resource files are stored.
+
+    Returns
+    -------
+    responses: list of dict
+        Responses as are loaded from the entry points.
+        Note that the content has not been sanitized nor validated.
+    """
+    request = {}
+    responses = []
+    for entry_point in pkg_resources.iter_entry_points(ENTRY_POINT_GROUP):
+        try:
+            loader = entry_point.load()
+        except ImportError:
+            logger.exception("Failed to load entry point %r", entry_point)
+            continue
+
+        try:
+            response = loader(request)
+        except Exception:
+            logger.exception("Failed to obtain data from %r", entry_point)
+            continue
+        responses.append(response)
+    return responses
+
+
+def response_to_node(response):
+    """ Convert a response (dict) to an instance of DemoTreeNodeObject
+
+    Parameters
+    ----------
+    response : dict
+        Response obtained from an entry point.
+
+    Returns
+    -------
+    node : DemoTreeNodeObject
+    """
+    root = response.get("root")
+    if root is not None and not os.path.exists(root):
+        return _get_placeholder_node(
+            response,
+            "Unable to load data: Path {!r} not found.".format(root)
+        )
+
+    try:
+        return DemoPath(
+            nice_name=response["name"],
+            name=response["root"],
+        )
+    except KeyError:
```

Comment:
Is this because `name` might not exist? Wouldn't it be better to move both of these checks (`root` and `name`) to the top of the function and raise the exception there?

Reply:
"name" may not exist, "root" may not exist, ... 

In fact, we can't even be sure that the entry point will return a `dict`. So actually I need to wrap this with another `try-except-log`.

This is to deal with data outside the control of `etsdemo`, so these defensive `try-except-log` are justified.